### PR TITLE
Xgboost tree tests

### LINKED ISSUE #3123: Refactor Xgboost TreeExplainer tests
I've previously refactored the `test_tree.py` file: Sklearn tree tests into its own TestClass (#3006), and LightGBM similarly (dsgibbons#88).

We should do the same for `XGBoost` as well. And ensure that Regressor/Classifier etc. use-cases are all covered.

# TODO

- [x] Rearrange tests into `TestExplainerXGBoost`.
- [x] Add proper tests for XGBRegressor (currently it's only tested indirectly via `test_xgboost_direct`). This is a follow-up task from #3121.

### PR DESCRIPTION
**Pull request by znacer**
## Overview

Closes #3123  

Description of the changes proposed in this pull request:
- [x] Rearrange tests into TestExplainerXGBoost  
- [x] Add proper tests for XGBRegressor  
- [ ] Complete FIXME in concerned tests (passing a `random_seed` argument)  
- [x] Fix a remaining warning due to an unused parameter.

## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

File: tests/explainers/test_tree.py [hunk 1]
```diff
@@ -889,8 +742,8 @@ def test_sum_match_random_forest_classifier(self):
 
         # check that SHAP values sum to model output
         assert (
-            np.abs(shap_values[0].sum(1) + ex.expected_value[0] - predicted[:, 0]).max()
-            < 1e-4
+                np.abs(shap_values[0].sum(1) + ex.expected_value[0] - predicted[:, 0]).max()
+                < 1e-4
```
Comment (thatlittleboy) [inline tests/explainers/test_tree.py]: Please revert all the formatting changes in `TestSingleTree` as it's not really relevant to this PR, and it should be 1 level of indentation instead of 2 anyway.

File: tests/explainers/test_tree.py [hunk 2]
```diff
@@ -1205,13 +1058,175 @@ def test_HistGradientBoostingRegressor(self):
...
+    Included models:
+        * XGBRegressor
+        * XGBClassifier
+    """
+
+    def test_xgboost(self):
+        """Test the basic `shap_values` calculation."""
+        xgboost = pytest.importorskip('xgboost')
+
+        # train lightgbm model
```
Comment (thatlittleboy) [inline tests/explainers/test_tree.py]: xgboost, not lightgbm

File: tests/explainers/test_tree.py [hunk 3]
```diff
@@ -1205,13 +1058,175 @@ def test_HistGradientBoostingRegressor(self):
...
+        predicted = model.predict(X)
+
+        expected_diff = np.abs(shap_values.sum(1) + ex.expected_value - predicted).max()
+        assert expected_diff < 1e-4, "SHAP values don't sum to model output!"
+
+    def test_xgboost_direct(self, random_seed):
+        xgboost = pytest.importorskip('xgboost')
+
+        rs = np.random.RandomState(random_seed)
+        N = 100
```
Comment (thatlittleboy) [inline tests/explainers/test_tree.py]: please reintroduce the FIXME about the random seed here. The issue is not resolved yet.

(try setting the random seed to 28 and the test will fail).

File: tests/explainers/test_tree.py [hunk 4]
```diff
@@ -1205,13 +1058,175 @@ def test_HistGradientBoostingRegressor(self):
...
+        X["HouseAge"] = X["HouseAge"].astype(np.int64)
+        X['IsOld'] = (X['HouseAge'] > 30)
+        bst = xgboost.train({"learning_rate": 0.01}, xgboost.DMatrix(X, label=y), 1000)
+        shap_values = shap.TreeExplainer(bst).shap_values(X)
+        shap.dependence_plot(0, shap_values, X, show=False)
+
+    def test_xgboost_classifier_independent_margin(self, random_seed):
+        xgboost = pytest.importorskip("xgboost")
+        # train XGBoost model
+        rs = np.random.RandomState(random_seed)
```
Comment (thatlittleboy) [inline tests/explainers/test_tree.py]: same comment here in `test_xgboost_classifier_independent_margin` regarding the random seed. Please retain the FIXME and remove `random_seed` for now.

File: tests/explainers/test_tree.py [hunk 5]
```diff
@@ -1205,13 +1058,175 @@ def test_HistGradientBoostingRegressor(self):
...
         )
 
 
+class TestExplainerXGBoost:
+    """Tests for the TreeExplainer with XGBoost models.
+
+    Included models:
+        * XGBRegressor
+        * XGBClassifier
+    """
```
Comment (thatlittleboy) [inline tests/explainers/test_tree.py]: XGBRanker is also tested in this TestClass

**Review by thatlittleboy — CHANGES_REQUESTED** [review]
Some minor changes, thanks for picking up this issue / PR!

File: tests/explainers/test_tree.py (hunk 1 above)
Comment (znacer) [reply tests/explainers/test_tree.py]: Corrected by commit [d0c5987](https://github.com/shap/shap/pull/3192/commits/d0c59873e0949bb20a8ac42a690d44ebb22b6f45)

File: tests/explainers/test_tree.py (hunk 2 above)
Comment (znacer) [reply tests/explainers/test_tree.py]: Corrected by commit [d0c5987](https://github.com/shap/shap/pull/3192/commits/d0c59873e0949bb20a8ac42a690d44ebb22b6f45)

File: tests/explainers/test_tree.py (hunk 3 above)
Comment (znacer) [reply tests/explainers/test_tree.py]: Commit reverted

File: tests/explainers/test_tree.py (hunk 4 above)
Comment (znacer) [reply tests/explainers/test_tree.py]: Commit reverted

File: tests/explainers/test_tree.py (hunk 5 above)
Comment (znacer) [reply tests/explainers/test_tree.py]: Corrected by commit [d0c5987](https://github.com/shap/shap/pull/3192/commits/d0c59873e0949bb20a8ac42a690d44ebb22b6f45)

**Review by thatlittleboy — APPROVED** [review]
fluffy-dog-air-swimming
